test/distributions/test_distributions.py

```diff
@@ -4665,6 +4665,87 @@ def test_mean(self):
             else:
                 self.assertEqual(pytorch_dist.mean, scipy_dist.mean(), msg=pytorch_dist)
 
+    def _test_discrete_distribution_mode(self, dist, sanitized_mode):
+        # We cannot easily check the mode for discrete distributions, but we can look left and right
+        # to ensure the log probability is smaller than at the mode.
+        isfinite = dist.mode.isfinite().reshape(dist.batch_shape + (dist.event_shape.numel(),)).all(-1)
+        for step in [-1, 1]:
+            log_prob_mode = dist.log_prob(sanitized_mode)
+            if isinstance(dist, OneHotCategorical):
+                idx = (dist._categorical.mode + 1) % dist.probs.shape[-1]
+                other = torch.nn.functional.one_hot(idx, num_classes=dist.probs.shape[-1]).to(dist.mode)
+            else:
+                other = dist.mode + step
+            mask = isfinite & dist.support.check(other)
+            self.assertTrue(mask.any() or dist.mode.unique().numel() == 1)
+            other = torch.where(mask, other, dist.sample())
+            log_prob_other = dist.log_prob(other)
+            np.testing.assert_array_less(log_prob_other[mask], log_prob_mode[mask])
+
+    def _test_continuous_distribution_mode(self, dist, sanitized_mode):
+        sanitized_mode = sanitized_mode.requires_grad_()
+        log_probs = dist.log_prob(sanitized_mode)
+        assert log_probs.shape == dist.batch_shape
+
+        # We evaluate the "boundary sign" which is -1 if the mode is on the left boundary of its
+        # support, +1 if on the right boundary, and 0 if not on the boundary.
+        if isinstance(dist.support, torch.distributions.constraints._GreaterThanEq):
+            boundary_sign = torch.where(dist.support.lower_bound == dist.mode, -1., 0.)
+        elif isinstance(dist.support, torch.distributions.constraints._Interval):
+            boundary_sign = torch.where(
+                dist.support.lower_bound == dist.mode, -1.,
+                torch.where(dist.support.upper_bound == dist.mode, 1., 0.)
+            )
+        else:
+            boundary_sign = torch.zeros_like(sanitized_mode)
```

Comment:
I don't trust boundary computations.  How confident are you that this code is foolproof and robust to numerical precision errors?